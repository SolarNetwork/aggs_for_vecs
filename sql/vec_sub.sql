-- int sub to scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(5::int, ints) FROM measurements WHERE sensor_id IN (4)) r(r);
-- int sub to scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(ints, 5::int) FROM measurements WHERE sensor_id IN (4)) r(r);
-- int sub to array
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(ints, ints) FROM measurements WHERE sensor_id IN (4)) r(r);
-- int sub to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::int, ints) FROM measurements WHERE sensor_id IN (4)) r(r);
-- int sub to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(ints, NULL::int) FROM measurements WHERE sensor_id IN (4)) r(r);
-- int sub scalar to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(5::int, NULL::int[]) FROM measurements WHERE sensor_id IN (4)) r(r);
-- int sub scalar to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::int[], 5::int) FROM measurements WHERE sensor_id IN (4)) r(r);
-- int sub to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(ints, NULL::int[]) FROM measurements WHERE sensor_id IN (4)) r(r);
-- int sub to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::int[], ints) FROM measurements WHERE sensor_id IN (4)) r(r);
-- null int sub to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::int, NULL::int[]) FROM measurements WHERE sensor_id IN (4)) r(r);
-- null int sub to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::int[], NULL::int) FROM measurements WHERE sensor_id IN (4)) r(r);
-- null int sub to null array
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::int[], NULL::int[]) FROM measurements WHERE sensor_id IN (4)) r(r);

-- float sub to scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(5::float, floats) FROM measurements WHERE sensor_id IN (4)) r(r);
-- float sub to scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(floats, 5::float) FROM measurements WHERE sensor_id IN (4)) r(r);
-- float sub to array
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(floats, floats) FROM measurements WHERE sensor_id IN (4)) r(r);
-- float sub to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::float, floats) FROM measurements WHERE sensor_id IN (4)) r(r);
-- float sub to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(floats, NULL::float) FROM measurements WHERE sensor_id IN (4)) r(r);
-- float sub scalar to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(5::float, NULL::float[]) FROM measurements WHERE sensor_id IN (4)) r(r);
-- float sub scalar to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::float[], 5::float) FROM measurements WHERE sensor_id IN (4)) r(r);
-- float sub to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(floats, NULL::float[]) FROM measurements WHERE sensor_id IN (4)) r(r);
-- float sub to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::float[], floats) FROM measurements WHERE sensor_id IN (4)) r(r);
-- null float sub to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::float, NULL::float[]) FROM measurements WHERE sensor_id IN (4)) r(r);
-- null float sub to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::float[], NULL::float) FROM measurements WHERE sensor_id IN (4)) r(r);
-- null float sub to null array
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::float[], NULL::float[]) FROM measurements WHERE sensor_id IN (4)) r(r);

-- numeric sub to scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(5::numeric, nums) FROM measurements WHERE sensor_id IN (4)) r(r);
-- numeric sub to scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(nums, 5::numeric) FROM measurements WHERE sensor_id IN (4)) r(r);
-- numeric sub to array
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(nums, nums) FROM measurements WHERE sensor_id IN (4)) r(r);
-- numeric sub to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::numeric, nums) FROM measurements WHERE sensor_id IN (4)) r(r);
-- numeric sub to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(nums, NULL::numeric) FROM measurements WHERE sensor_id IN (4)) r(r);
-- numeric sub scalar to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(5::numeric, NULL::numeric[]) FROM measurements WHERE sensor_id IN (4)) r(r);
-- numeric sub scalar to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::numeric[], 5::numeric) FROM measurements WHERE sensor_id IN (4)) r(r);
-- numeric sub to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(nums, NULL::numeric[]) FROM measurements WHERE sensor_id IN (4)) r(r);
-- numeric sub to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::numeric[], nums) FROM measurements WHERE sensor_id IN (4)) r(r);
-- null numeric sub to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::numeric, NULL::numeric[]) FROM measurements WHERE sensor_id IN (4)) r(r);
-- null numeric sub to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::numeric[], NULL::numeric) FROM measurements WHERE sensor_id IN (4)) r(r);
-- null numeric sub to null array
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::numeric[], NULL::numeric[]) FROM measurements WHERE sensor_id IN (4)) r(r);
