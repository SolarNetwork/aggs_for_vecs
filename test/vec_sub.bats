load test_helper

@test "int16 sub" {
  result="$(query "SELECT vec_sub(ARRAY[1,2,3]::smallint[], ARRAY[3,2,1]::smallint[])")";
  echo $result;
  [ "$result" = "{-2,0,2}" ]
}

@test "int32 sub" {
  result="$(query "SELECT vec_sub(ARRAY[1,2,3]::integer[], ARRAY[3,2,1]::integer[])")";
  echo $result;
  [ "$result" = "{-2,0,2}" ]
}

@test "int64 sub" {
  result="$(query "SELECT vec_sub(ARRAY[1,2,3]::bigint[], ARRAY[3,2,1]::bigint[])")";
  echo $result;
  [ "$result" = "{-2,0,2}" ]
}

@test "float32 sub" {
  result="$(query "SELECT vec_sub(ARRAY[1.0,2.2,3.4]::real[], ARRAY[3.4,2.2,1.0]::real[])")";
  echo $result;
  [ "$result" = "{-2.4,0,2.4}" ]
}

@test "float64 sub" {
  result="$(query "SELECT vec_sub(ARRAY[1.1,2.2,3.4]::float[], ARRAY[3.4,2.2,1.1]::float[])")";
  echo $result;
  [ "$result" = "{-2.3,0,2.3}" ]
}

@test "numeric sub" {
  result="$(query "SELECT vec_sub(ARRAY[1.23,2.234,3.456]::numeric[], ARRAY[3.456,2.234,1.123]::numeric[])")";
  echo $result;
  [ "$result" = "{-2.226,0.000,2.333}" ]
}

@test "numeric sub measurements" {
  result="$(query "SELECT diff FROM (SELECT vec_sub(nums, lag(nums) OVER (ORDER BY sensor_id)) AS diff FROM measurements d WHERE sensor_id IN (3,4)) d WHERE d.diff IS NOT NULL")";
  echo $result;
  [ "$result" = "{0.00,NULL,-1.11}" ]
}
