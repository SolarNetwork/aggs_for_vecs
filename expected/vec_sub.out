-- int sub to scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(5::int, ints) FROM measurements WHERE sensor_id IN (4)) r(r);
     r      | pg_typeof 
------------+-----------
 {4,NULL,3} | integer[]
(1 row)

-- int sub to scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(ints, 5::int) FROM measurements WHERE sensor_id IN (4)) r(r);
      r       | pg_typeof 
--------------+-----------
 {-4,NULL,-3} | integer[]
(1 row)

-- int sub to array
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(ints, ints) FROM measurements WHERE sensor_id IN (4)) r(r);
     r      | pg_typeof 
------------+-----------
 {0,NULL,0} | integer[]
(1 row)

-- int sub to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::int, ints) FROM measurements WHERE sensor_id IN (4)) r(r);
        r         | pg_typeof 
------------------+-----------
 {NULL,NULL,NULL} | integer[]
(1 row)

-- int sub to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(ints, NULL::int) FROM measurements WHERE sensor_id IN (4)) r(r);
        r         | pg_typeof 
------------------+-----------
 {NULL,NULL,NULL} | integer[]
(1 row)

-- int sub scalar to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(5::int, NULL::int[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | integer[]
(1 row)

-- int sub scalar to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::int[], 5::int) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | integer[]
(1 row)

-- int sub to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(ints, NULL::int[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | integer[]
(1 row)

-- int sub to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::int[], ints) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | integer[]
(1 row)

-- null int sub to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::int, NULL::int[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | integer[]
(1 row)

-- null int sub to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::int[], NULL::int) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | integer[]
(1 row)

-- null int sub to null array
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::int[], NULL::int[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | integer[]
(1 row)

-- float sub to scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(5::float, floats) FROM measurements WHERE sensor_id IN (4)) r(r);
     r      |     pg_typeof      
------------+--------------------
 {4,NULL,3} | double precision[]
(1 row)

-- float sub to scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(floats, 5::float) FROM measurements WHERE sensor_id IN (4)) r(r);
      r       |     pg_typeof      
--------------+--------------------
 {-4,NULL,-3} | double precision[]
(1 row)

-- float sub to array
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(floats, floats) FROM measurements WHERE sensor_id IN (4)) r(r);
     r      |     pg_typeof      
------------+--------------------
 {0,NULL,0} | double precision[]
(1 row)

-- float sub to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::float, floats) FROM measurements WHERE sensor_id IN (4)) r(r);
        r         |     pg_typeof      
------------------+--------------------
 {NULL,NULL,NULL} | double precision[]
(1 row)

-- float sub to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(floats, NULL::float) FROM measurements WHERE sensor_id IN (4)) r(r);
        r         |     pg_typeof      
------------------+--------------------
 {NULL,NULL,NULL} | double precision[]
(1 row)

-- float sub scalar to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(5::float, NULL::float[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r |     pg_typeof      
---+--------------------
   | double precision[]
(1 row)

-- float sub scalar to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::float[], 5::float) FROM measurements WHERE sensor_id IN (4)) r(r);
 r |     pg_typeof      
---+--------------------
   | double precision[]
(1 row)

-- float sub to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(floats, NULL::float[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r |     pg_typeof      
---+--------------------
   | double precision[]
(1 row)

-- float sub to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::float[], floats) FROM measurements WHERE sensor_id IN (4)) r(r);
 r |     pg_typeof      
---+--------------------
   | double precision[]
(1 row)

-- null float sub to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::float, NULL::float[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r |     pg_typeof      
---+--------------------
   | double precision[]
(1 row)

-- null float sub to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::float[], NULL::float) FROM measurements WHERE sensor_id IN (4)) r(r);
 r |     pg_typeof      
---+--------------------
   | double precision[]
(1 row)

-- null float sub to null array
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::float[], NULL::float[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r |     pg_typeof      
---+--------------------
   | double precision[]
(1 row)

-- numeric sub to scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(5::numeric, nums) FROM measurements WHERE sensor_id IN (4)) r(r);
        r         | pg_typeof 
------------------+-----------
 {3.77,NULL,2.66} | numeric[]
(1 row)

-- numeric sub to scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(nums, 5::numeric) FROM measurements WHERE sensor_id IN (4)) r(r);
         r          | pg_typeof 
--------------------+-----------
 {-3.77,NULL,-2.66} | numeric[]
(1 row)

-- numeric sub to array
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(nums, nums) FROM measurements WHERE sensor_id IN (4)) r(r);
        r         | pg_typeof 
------------------+-----------
 {0.00,NULL,0.00} | numeric[]
(1 row)

-- numeric sub to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::numeric, nums) FROM measurements WHERE sensor_id IN (4)) r(r);
        r         | pg_typeof 
------------------+-----------
 {NULL,NULL,NULL} | numeric[]
(1 row)

-- numeric sub to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(nums, NULL::numeric) FROM measurements WHERE sensor_id IN (4)) r(r);
        r         | pg_typeof 
------------------+-----------
 {NULL,NULL,NULL} | numeric[]
(1 row)

-- numeric sub scalar to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(5::numeric, NULL::numeric[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | numeric[]
(1 row)

-- numeric sub scalar to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::numeric[], 5::numeric) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | numeric[]
(1 row)

-- numeric sub to null array lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(nums, NULL::numeric[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | numeric[]
(1 row)

-- numeric sub to null array rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::numeric[], nums) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | numeric[]
(1 row)

-- null numeric sub to null scalar lhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::numeric, NULL::numeric[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | numeric[]
(1 row)

-- null numeric sub to null scalar rhs
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::numeric[], NULL::numeric) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | numeric[]
(1 row)

-- null numeric sub to null array
SELECT r, pg_typeof(r) FROM (SELECT vec_sub(NULL::numeric[], NULL::numeric[]) FROM measurements WHERE sensor_id IN (4)) r(r);
 r | pg_typeof 
---+-----------
   | numeric[]
(1 row)

